appengine/genetics/js/cage.js

```diff
@@ -0,0 +1,730 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview JavaScript for the game logic for the Genetics game.
+ * @author kozbial@google.com (Monica Kozbial)
+ */
+'use strict';
+
+goog.provide('Genetics.Cage');
+
+goog.require('Genetics.Mouse');
+
+/**
+ * The maximum population in a cage.
+ * @type {number}
+ */
+Genetics.Cage.MAX_POPULATION = 10;
+
+/**
+ * The number of mice added to the cage for each player at the start of the
+ * game. Should be less than MAX_POPULATION * number of players.
+ * @type {number}
+ */
+Genetics.Cage.START_MICE_PER_PLAYER = 2;
+
+/**
+ * Mapping of player id to and array with player name and code.
+ * @type {!Object.<number, Array.<string>>}
+ */
+Genetics.Cage.PLAYERS = {};
+
+/**
+ * Whether the simulation is running without a visualization.
+ * @type {boolean}
+ */
+Genetics.Cage.HEADLESS = false;
+
+/**
+ * List of mice currently alive.
+ * @type {!Array.<Genetics.Mouse>}
+ * @private
+ */
+Genetics.Cage.aliveMice_ = [];
+
+/**
+ * Mapping of mouse id to mouse for quick lookup.
+ * @type {!Object.<number, Genetics.Mouse>}
+ * @private
+ */
+Genetics.Cage.miceMap_ = {};
+
+/**
+ * Mapping of mouse id to PID of life simulation for that mouse queued to run.
+ * @type {!Object.<number, number>}
+ * @private
+ */
+Genetics.Cage.mouseLife_ = {};
+
+/**
+ * List of events to be visualized.
+ * @type {!Array.<!Event>}
+ */
+Genetics.Cage.EVENTS = [];
+
+/**
+ * Template for event properties.
+ * @type {Object.<string, Array.<string>>}
+ */
+Genetics.Cage.EVENT_TEMPLATE = {
+  'ADD': ['TYPE', 'MOUSE', 'PLAYER_NAME'],
+  'START_GAME': [],
+  'FIGHT': ['TYPE', 'ID', 'RESULT', 'OPT_OPPONENT'],
+  'MATE': ['TYPE', 'ID', 'RESULT', 'OPT_PARTNER', 'OPT_OFFSPRING'],
+  // A mouse dies after using all fight and mate opportunities.
+  'RETIRE': ['TYPE', 'ID'],
+  // The oldest mouse dies when there is no room for it.
+  'OVERPOPULATION': ['TYPE', 'ID'],
+  // A mouse dies if it throws an exception.
+  'EXPLODE': ['TYPE', 'ID', 'SOURCE', 'OPT_CAUSE'],
+  // A mouse dies if it does not complete execution.
+  'SPIN': ['TYPE', 'ID'],
+  'END_GAME': ['TYPE', 'CAUSE', 'PICK_FIGHT_WINNER', 'CHOOSE_MATE_WINNER',
+    'MATE_ANSWER_WINNER']
+};
+
+/**
+ * Creates an event.
+ * @param {string } type The type of event.
+ * @param {...string|number|Genetics.Mouse} var_args The properties on
+ * the event.
+ * @constructor
+ */
+Genetics.Cage.Event = function(type, var_args) {
+  var template = Genetics.Cage.EVENT_TEMPLATE[type];
```

Comment:
I'm not modifying EVENT_TEMPLATE. I'm using type of the event to get a list of event properties from EVENT_TEMPLATE and using these properties as keys on the event object that map to the arguments passed in the constructor.